### ISSUE TITLE
Support for spark in yarn-client or yarn-cluster modes?

### ISSUE THREAD

**Issue by TheLoneRangerShack** [NFLabs/z-manager#11]
The multitenancy README page says that only the standalone mode is supported for Spark. Unfortunately, we cannot use standalone mode in our cluster, and can only work with yarn at the moment. Is yarn support coming, and if so, when?

**Comment by khalidhuseynov** [comment]
You're right, currently we do not support the yarn mode. It's definitely part of scope for future improvements on Z-Manager multitenancy, but not scheduled for execution in the coming 1-2 months. For more details you can also refer to the thread  [here](http://mail-archives.apache.org/mod_mbox/incubator-zeppelin-users/201506.mbox/%3CCAKOFcwpkfntaoJWony%3DjndA_TbDo%2BLH3symRpzAgKN3%2Bd_BOgw%40mail.gmail.com%3E) on the user mailing list about Zeppelin in multi-user environment. Let us know if you have any other questions.

**Comment by TheLoneRangerShack** [comment]
Thanks for the info! While the thread you linked to has helped me understand why multiple users using a single instance of Zeppelin is problematic, I'm still not clear why z-manager can't work with yarn mode. If you could point to some documentation explaining this, or if you could clarify it yourself, I would appreciate it!

**Comment by bzz** [comment]
@TheLoneRangerShack thank your interest in z-manager! 

@khalidhuseynov did a great job highlighting the current situation, now let me try to answer your question by digging a bit deeper into the details.

**Z-Manager multi-tenancy** (not to be confused with second part of **Z-Manager Installer** - a .sh script that just downloads Z to local machine) is fundamentally a reverse-proxy, it spins up a new Docker container per-user and proxies user access to that particular instance of Zeppelin, running inside the container.

Now the tricky part is - Zeppelin starts a Spark Driver inside the container and it should be [accessable from\to spark master and workers](http://spark.apache.org/docs/latest/security.html#configuring-ports-for-network-security). In Spark Standalone mode - it just requires 
- a) networking configuration of the container - setting the port forwarding for the [number of known ports](http://spark.apache.org/docs/latest/configuration.html#networking) for spark sub-systems (classloader server, file server, etc) and 
- b) Spark configuration on the zeppelin side (thought the SparkConf, before context creating), to make sure Spark indeed uses these [number of known ports](https://github.com/NFLabs/z-manager/blob/master/multitenancy/server/docker.go#L54) for communication, that are already open and forwarded betwee the container and host machine.

So one can imagine that in YARN mode  (zeppelin supports only yarn-client for now) - there will be different b) part, which needs to be taken care of separately and that was the plan. Right now we do not have use-case for that at hand, but it might come up in next few months.

That been said, contributions are very welcome and please do not hesitate to ask here or on the [Mailing List](https://groups.google.com/forum/#!forum/z-manager) if you have any further question on getting z-manager running, adding YARN support, etc.

**Comment by b3nbk1m70** [comment]
YARN client mode is fine. Can z-manager support this mode? Cluster mode is not necessary in most circumstances where stdout is needed.
